sgl-kernel/csrc/moe/moe_sum_reduce.cu

```diff
@@ -0,0 +1,315 @@
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <cuda.h>
+#include <cudaTypedefs.h>
+#include <cuda_runtime.h>
+#include <torch/all.h>
+
+#include <iostream>
+#include <type_traits>
+
+#include "cutlass/array.h"
+#include "utils.h"
+
+template <typename T>
+__device__ __forceinline__ float to_float(T x) {
+  return static_cast<float>(x);
+}
+
+template <>
+__device__ __forceinline__ float to_float<half>(half x) {
+  return __half2float(x);
+}
+
+template <typename T>
+__device__ __forceinline__ T from_float(float x) {
+  return static_cast<T>(x);
+}
+
+template <>
+__device__ __forceinline__ half from_float<half>(float x) {
+  return __float2half_rn(x);
+}
+
+template <typename T>
+__device__ __forceinline__ T ldg_cg(const T* p) {
+  return __ldg(p);
+}
+
+__device__ __forceinline__ float bf16_to_float_bits(uint16_t h) {
+  uint32_t u = ((uint32_t)h) << 16;
+  return __uint_as_float(u);
+}
+__device__ __forceinline__ uint16_t float_to_bf16_rne(float x) {
+  uint32_t u = __float_as_uint(x);
+  uint32_t lsb = (u >> 16) & 1u;
+  uint32_t rounding_bias = 0x7FFFu + lsb;
+  return (uint16_t)((u + rounding_bias) >> 16);
+}
+
+union Pack16B {
+  uint4 v;
+  uint16_t u16[8];
+};
+
+template <int WARPS_PER_BLOCK, int VEC = 16>
+__global__ void moe_sum_reduce_kernel_warp_token_topk_bf16_vec(
+    const at::BFloat16* __restrict__ x,
+    at::BFloat16* __restrict__ y,
+    const int64_t token_num,
+    const int64_t hidden_dim,
+    const int64_t topk_num,
+    const int64_t stride_token,      // in elements
+    const int64_t stride_topk,       // in elements
+    const int64_t out_stride_token,  // in elements
+    const float scale) {
+  static_assert(VEC % 8 == 0, "VEC must be a multiple of 8");
+  constexpr int PACKS = VEC / 8;
+
+  const int warp_id = threadIdx.x / 32;
+  const int lane = threadIdx.x % 32;
+  const int64_t t = (int64_t)blockIdx.y * WARPS_PER_BLOCK + warp_id;
+  if (t >= token_num) return;
+
+  const int64_t n_chunks = hidden_dim / VEC;
+
+  for (int64_t chunk = (int64_t)blockIdx.x * 32 + lane; chunk < n_chunks; chunk += (int64_t)gridDim.x * 32) {
+    const int64_t d = chunk * VEC;
+    const int64_t base = t * stride_token + d;
+
+    float acc[VEC];
+#pragma unroll
+    for (int i = 0; i < VEC; ++i)
+      acc[i] = 0.f;
+
+#pragma unroll
+    for (int k = 0; k < topk_num; ++k) {
+#pragma unroll
+      for (int p = 0; p < PACKS; ++p) {
+        const int64_t offset = base + (int64_t)k * stride_topk + p * 8;
+
+        Pack16B pack = {ldg_cg(reinterpret_cast<const uint4*>(x + offset))};
```

Comment:
Should verify that `(x + offset)` is 16-byte aligned for uint4 access?

Reply:
input x is bf16 make sure the data is 16-bytes aligned for uint4 access.
Adding the check in loop increase the latency.
```
#pragma unroll
    for (int k = 0; k < topk_num; ++k) {
#pragma unroll
      for (int p = 0; p < PACKS; ++p) {
        const int64_t offset = base + (int64_t)k * stride_topk + p * 8;

        TORCH_CHECK(reinterpret_cast<uintptr_t>(x + offset) % 16 == 0,
            "Memory is not 16-byte aligned for uint4 access");
        Pack16B pack = {ldg_cg(reinterpret_cast<const uint4*>(x + offset))};
```